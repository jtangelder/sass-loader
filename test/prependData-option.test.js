import nodeSass from 'node-sass';
import dartSass from 'sass';
import Fiber from 'fibers';

import {
  compile,
  getCodeFromBundle,
  getCodeFromSass,
  getCompiler,
  getErrors,
  getImplementationByName,
  getTestId,
  getWarnings,
} from './helpers';

const implementations = [nodeSass, dartSass];
const syntaxStyles = ['scss', 'sass'];

describe('prependData option', () => {
  beforeEach(() => {
    // The `sass` (`Dart Sass`) package modify the `Function` prototype, but the `jest` lose a prototype
    Object.setPrototypeOf(Fiber, Function.prototype);
  });

  implementations.forEach((implementation) => {
    const [implementationName] = implementation.info.split('\t');

    syntaxStyles.forEach((syntax) => {
      it(`should work with the "data" option as a string (${implementationName}) (${syntax})`, async () => {
        const testId = getTestId('prepending-data', syntax);
        const options = {
          implementation: getImplementationByName(implementationName),
          prependData: `$prepended-data: hotpink${
            syntax === 'sass' ? '\n' : ';'
          }`,
        };
        const compiler = getCompiler(testId, { loader: { options } });
        const stats = await compile(compiler);
        const codeFromBundle = getCodeFromBundle(stats, compiler);
        const codeFromSass = getCodeFromSass(testId, options);

        expect(codeFromBundle.css).toBe(codeFromSass.css);
        expect(codeFromBundle.css).toMatchSnapshot('css');
        expect(getWarnings(stats)).toMatchSnapshot('warnings');
        expect(getErrors(stats)).toMatchSnapshot('errors');
      });

      it(`should work with the "data" option as a function (${implementationName}) (${syntax})`, async () => {
        const testId = getTestId('prepending-data', syntax);
        const options = {
          implementation: getImplementationByName(implementationName),
          prependData: (loaderContext) => {
            expect(loaderContext).toBeDefined();

            return `$prepended-data: hotpink${syntax === 'sass' ? '\n' : ';'}`;
          },
        };
        const compiler = getCompiler(testId, { loader: { options } });
        const stats = await compile(compiler);
        const codeFromBundle = getCodeFromBundle(stats, compiler);
        const codeFromSass = getCodeFromSass(testId, options);

        expect(codeFromBundle.css).toBe(codeFromSass.css);
        expect(codeFromBundle.css).toMatchSnapshot('css');
        expect(getWarnings(stats)).toMatchSnapshot('warnings');
        expect(getErrors(stats)).toMatchSnapshot('errors');
      });

      it(`should use same EOL on all os (${implementationName}) (${syntax})`, async () => {
        const testId = getTestId('prepending-data', syntax);
        const prependData =
          syntax === 'sass'
            ? `$prepended-data: hotpink
a
  color: $prepended-data`
            : `$prepended-data: hotpink;
a {
  color: red;
}`;
        const options = {
          implementation: getImplementationByName(implementationName),
          prependData,
        };
        const compiler = getCompiler(testId, { loader: { options } });
        const stats = await compile(compiler);
        const codeFromBundle = getCodeFromBundle(stats, compiler);

        expect(codeFromBundle.css).toMatchSnapshot('css');
        expect(getWarnings(stats)).toMatchSnapshot('warnings');
        expect(getErrors(stats)).toMatchSnapshot('errors');
      });
    });
  });
});
