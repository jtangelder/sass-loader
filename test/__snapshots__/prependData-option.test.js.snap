// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prependData option should work with the "data" option as a function (dart-sass) (sass): errors 1`] = `Array []`;

exports[`prependData option should work with the "data" option as a function (dart-sass) (sass): warnings 1`] = `Array []`;

exports[`prependData option should work with the "data" option as a function (dart-sass) (scss): errors 1`] = `Array []`;

exports[`prependData option should work with the "data" option as a function (dart-sass) (scss): warnings 1`] = `Array []`;

exports[`prependData option should work with the "data" option as a function (node-sass) (sass): errors 1`] = `Array []`;

exports[`prependData option should work with the "data" option as a function (node-sass) (sass): warnings 1`] = `Array []`;

exports[`prependData option should work with the "data" option as a function (node-sass) (scss): errors 1`] = `Array []`;

exports[`prependData option should work with the "data" option as a function (node-sass) (scss): warnings 1`] = `Array []`;

exports[`prependData option should work with the "data" option as a string (dart-sass) (sass): errors 1`] = `Array []`;

exports[`prependData option should work with the "data" option as a string (dart-sass) (sass): warnings 1`] = `Array []`;

exports[`prependData option should work with the "data" option as a string (dart-sass) (scss): errors 1`] = `Array []`;

exports[`prependData option should work with the "data" option as a string (dart-sass) (scss): warnings 1`] = `Array []`;

exports[`prependData option should work with the "data" option as a string (node-sass) (sass): errors 1`] = `Array []`;

exports[`prependData option should work with the "data" option as a string (node-sass) (sass): warnings 1`] = `Array []`;

exports[`prependData option should work with the "data" option as a string (node-sass) (scss): errors 1`] = `Array []`;

exports[`prependData option should work with the "data" option as a string (node-sass) (scss): warnings 1`] = `Array []`;
